src/postprocess/chi0.jl

```diff
@@ -103,33 +103,101 @@ LinearAlgebra.ldiv!(y::T, P::FunctionPreconditioner, x) where {T} = P.preconditi
 LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(similar(x), x))
 precondprep!(P::FunctionPreconditioner, ::Any) = P
 
-# Solves Q (H-εn) Q δψn = -Q rhs
-# where Q is the projector on the orthogonal of ψk
-function sternheimer_solver(Hk, ψk, ψnk, εnk, rhs; tol_cg=1e-6, verbose=false)
+# Solves (1-P) (H-εn) (1-P) δψn = - (1-P) rhs
+# where 1-P is the projector on the orthogonal of ψk
+function sternheimer_solver(Hk, ψk, ψnk, εnk, rhs;
+                            ψk_extra=zeros(size(ψk,1), 0),
+                            εk_extra=zeros(0), schur=true,
+                            tol_cg=1e-12, verbose=false)
     basis = Hk.basis
     kpoint = Hk.kpoint
+    model = basis.model
+    temperature = model.temperature
+
+    # We use a Schur decomposition of the orthogonal of the occupied states
+    # where we still know some information from the partially converged
+    # nonoccupied states (in particular, they are Rayleigh-Ritz wrt to Hk).
+    #
+    # /!\ This is only implemented for insulators at the moment, WIP for
+    # metals, in which case we use empty arrays and all computations are
+    # equivalent to invert the Sternheimer system in Ran(1-P) without the Schur
+    # trick.
+
+    # TODO
+    # finite temperature not supported yet, so we remove extra bands and perform
+    # all the computations with empty arrays
+    if !iszero(temperature) || (schur && (isempty(ψk_extra) || isempty(εk_extra)))
+        ψk_extra = zeros(size(ψk,1), 0)
+        εk_extra = zeros(0)
+        schur = false
+    end
 
-    # we err on the side of caution here by applying Q *a lot*
-    # there are optimizations to be made here
-    Q(ϕ) = ϕ - ψk * (ψk' * ϕ)
-    function QHQ(ϕ)
-        Qϕ = Q(ϕ)
-        Q(Hk * Qϕ - εnk * Qϕ)
+    # Projectors:
+    # projector onto the computed and converged states, which are the occupied
+    # states in the case of insulators
```

Comment:
They are also "occupied" for metals (in the sense of non-zero occupation). I would remove that half-sentence.